gordon/main.py

```diff
@@ -102,74 +102,41 @@ def _log_or_exit_on_exceptions(base_msg, exc, debug):
         raise SystemExit(1)
 
 
-def _assert_required_plugins(installed_plugins, debug):
-    # enricher not required
-    required_providers_available = {
-        'event_consumer': False,
-        'publisher': False,
-    }
-    for plugin in installed_plugins:
-        if interfaces.IEventConsumerClient.providedBy(plugin):
-            required_providers_available['event_consumer'] = True
-        elif interfaces.IPublisherClient.providedBy(plugin):
-            required_providers_available['publisher'] = True
-
-    missing = []
-    msg = ('The provider for the "{name}" interface is not configured for the '
-           'Gordon service or is not implemented.')
-    for provider, available in required_providers_available.items():
-        if not available:
-            exc = exceptions.MissingPluginError(msg.format(name=provider))
-            missing.append(exc)
-    if missing:
-        base_msg = 'Problem running plugins: '
-        _log_or_exit_on_exceptions(base_msg, missing, debug=debug)
-
-
 def _gather_runnable_plugins(plugins, debug):
     plugins_to_run = []
     for plugin in plugins:
-        if interfaces.IEventConsumerClient.providedBy(plugin):
+        if interfaces.IRunnable.providedBy(plugin):
             # TODO (lynn): this should be switched out for adding
             # the "verify interface implementation" ability. See
             # https://docs.zope.org/zope.interface/verify.html
             if not hasattr(plugin, 'run') or \
                     not asyncio.iscoroutinefunction(plugin.run):
                 msg = (f'Implemention "{plugin}" of the required '
-                       '"IEventConsumerClient" interface does not have the '
+                       '"IRunnable" interface does not have the '
                        'necessary `run` method.')
                 exc = exceptions.InvalidPluginError(msg)
                 _log_or_exit_on_exceptions(msg, exc, debug)
                 continue
             plugins_to_run.append(plugin)
-        elif hasattr(plugin, 'run') and asyncio.iscoroutinefunction(plugin.run):
-            plugins_to_run.append(plugin)
 
     return plugins_to_run
 
 
-def _gather_implemented_providers(plugins):
-    implemented = {}
-    for plugin in plugins:
-        if interfaces.IEventConsumerClient.providedBy(plugin):
-            implemented['event_consumer'] = plugin
-        elif interfaces.IEnricherClient.providedBy(plugin):
-            implemented['enricher'] = plugin
-        elif interfaces.IPublisherClient.providedBy(plugin):
-            implemented['publisher'] = plugin
-    return implemented
+def _assert_required_plugins(plugins, debug):
+    if not plugins:
+        msg = (f'At least one runnable plugin is required.')
```

Comment:
Should we require only a runnable one? or also one that implements a message handler so that it can be routed somewhere? The begs the question of "should we force plugin implementers/operators to have more than one plugin" since while I can foresee a use case (that I can't think of for now) that only uses one plugin, I'm not sure if the router would work without at least two it its current state.

Reply:
True, in that case it's not that it requires a message handler, but it requires a message handler that has phase 'cleanup', since that's where it would buckle. I'd almost rather just change the router, either to make final stages optional or have a default drop functionality for 'cleanup'. Alternatively, I could just add a check for 'cleanup' here. I'm slightly against requiring specific phases, so I think I like having a default for 'cleanup' best. Thoughts?